lib/Runtime/Language/Arguments.h

```diff
@@ -4,17 +4,58 @@
 //-------------------------------------------------------------------------------------------------------
 #pragma once
 
+// To extract variadic args array after known args list:
+//      argx, callInfo, ...
+// NOTE: The last known arg name is hard-coded to "callInfo".
 #ifdef _WIN32
-#define VA_LIST_TO_VARARRAY(vl, va, callInfo) Js::Var* va = (Js::Var*) vl;
+#define DECLARE_ARGS_VARARRAY(va, ...)                              \
+    va_list _vl;                                                    \
+    va_start(_vl, callInfo);                                        \
+    Js::Var* va = (Js::Var*)_vl
 #else
 #if defined(_M_X64) || defined(_M_IX86)
 // We use a custom calling convention to invoke JavascriptMethod based on
 // System V AMD64 ABI. At entry of JavascriptMethod the stack layout is:
 //      [Return Address] [function] [callInfo] [arg0] [arg1] ...
 //
-#define VA_LIST_TO_VARARRAY(vl, va, callInfo)                              \
-    Js::Var* va = reinterpret_cast<Js::Var*>(_AddressOfReturnAddress()) + 3; \
-    Assert(*reinterpret_cast<Js::CallInfo*>(va - 1) == callInfo);
+#define DECLARE_ARGS_VARARRAY_N(va, n)                              \
+    Js::Var* va = _get_va(_AddressOfReturnAddress(), n);            \
+    Assert(*reinterpret_cast<Js::CallInfo*>(va - 1) == callInfo)
+
+#define DECLARE_ARGS_VARARRAY(va, ...)                              \
+    DECLARE_ARGS_VARARRAY_N(va, _count_args(__VA_ARGS__))
+
+inline Js::Var* _get_va(void* addrOfReturnAddress, int n)
+{
+    Js::Var* pArgs = reinterpret_cast<Js::Var*>(addrOfReturnAddress) + 1;
```

Comment:
Oh right, my bad.

---
In reply to: [81249310](https://github.com/Microsoft/ChakraCore/pull/1591#discussion_r81249310) [](ancestors = 81249310)